pgrx/src/list.rs

```diff
@@ -12,257 +12,617 @@
+            assert!(drain_start + tail_len <= ffi::c_int::MAX as _);
+
+            // If draining all, rip it out of place to contain broken invariants from panics
+            let raw = if drain_start == 0 {
+                mem::take(self).into_nullable()
+            } else {
+                // Leave it in place, but we need a pointer:
+                match self {
+                    List::Nil => ptr::null_mut(),
+                    List::Cons(head) => head.list.as_ptr().cast(),
+                }
+            };
+
+            // Remember to check that our raw ptr is non-null
+            if raw != ptr::null_mut() {
+                // Shorten the list to prohibit interaction with List's state after drain_start.
+                // Note this breaks List repr invariants in the `drain_start == 0` case, but
+                // we only consider returning the list ptr to `&mut self` if Drop is completed
+                unsafe { (*raw).length = drain_start as _ };
+                let cells_ptr = unsafe { (*raw).elements };
+                let iter = unsafe {
+                    RawCellIter {
+                        ptr: cells_ptr.add(drain_start).cast(),
+                        end: cells_ptr.add(tail_start).cast(),
+                    }
+                };
+                Drain {
+                    tail_len: tail_len as _,
+                    tail_start: tail_start as _,
+                    raw,
+                    origin: self,
+                    iter,
+                }
+            } else {
+                // If it's not, produce the only valid choice: a 0-len iterator pointing to null
+                // One last doublecheck for old paranoia's sake:
+                assert!(tail_len == 0 && tail_start == 0 && drain_start == 0);
+                Drain { tail_len: 0, tail_start: 0, raw, origin: self, iter: Default::default() }
+            }
+        }
+
+        pub fn iter(&self) -> impl Iterator<Item = &T> {
+            self.as_cells().into_iter().map(Deref::deref)
+        }
+
+        pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> {
+            self.as_cells_mut().into_iter().map(DerefMut::deref_mut)
         }
     }
 
-    #[cfg(any(feature = "pg13", feature = "pg14", feature = "pg15", feature = "pg16"))]
-    #[inline]
-    pub fn replace_int(&mut self, i: usize, with: i32) {
-        unsafe {
-            let cell = pg_sys::pgrx_list_nth_cell(self.list, i as i32);
-            cell.as_mut().expect("cell is null").int_value = with;
+    impl<T> List<T> {
+        #[inline]
+        pub fn len(&self) -> usize {
+            match self {
+                List::Nil => 0,
+                List::Cons(head) => head.len(),
+            }
+        }
+
+        #[inline]
+        pub fn capacity(&self) -> usize {
+            match self {
+                List::Nil => 0,
+                List::Cons(head) => head.capacity(),
+            }
+        }
+
+        pub fn into_nullable(self) -> *mut pg_sys::List {
```

Comment:
I don't understand "nullable" here.  At the end of the day, this is returning the underlying list pointer.

I was looking for the function to "get the raw List pointer so it can be passed off to Postgres" and this looks like function, but the name doesn't make sense to me.

Reply:
It's a nullable i.e. maybe-null pointer.

In the rest of pgrx, we often rely on Postgres returning non-null pointers, and this one never is.

Reply:
pgrx has historically used `into_pg` for this and `from_pg` for `downcast_from_nullable`.  I guess that's what I was looking for.

I'm not arguing or anything.  This function name and also `downcast_from_nullable` just aren't what I was expecting.

Reply:
I could use `downcast_from_pg` but it's a "true" downcast in the sense that it's fallible.

Reply:
In practice, pgrx has used inconsistent names in a few places, e.g. https://github.com/pgcentralfoundation/pgrx/blob/047b1d1fc9e9c4007c871e226fa81e294f8bf5e6/pgrx/src/datum/array.rs#L161

and in my head it adopts an opposite polarity as e.g. [`as_non_null_ptr`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_non_null_ptr).

To me, the `from_pg` and `into_pg` names imply strongly the type is actually *from Postgres* or *going into Postgres*, but in actuality we have people composing usage of pgrx with *other active C code that also binds against Postgres*. This makes me uneasy with the idea that this other miscellaneous C code, which may also deal in Postgres types, should be treated as "inside the trust boundary" in the same way that Postgres itself *necessarily* is, and uneasy with names that may undermine someone's caution.

This, all together (alongside my gradually crumbling sanity as I tried to understand Postgres) is why I prioritized my personal taste when initially opening the PR. But, despite all those factors in tension, I don't necessarily feel married to their current names, especially if they feel non-intuitive.

Reply:
I don't want us to get bogged down in naming, especially as you point out, we're pretty inconsistent.  So I'll let you decide what you think is right and would be most expected by a random Rust programmer.  I didn't find these names obvious, but my opinions are biased.